### PR TITLE
Paypal notification handler

### PR DESCRIPTION
**Pull request by KaiNissen**

### CODE REVIEW

File: src/Infrastructure/PayPalPaymentNotificationVerifier.php [hunk 1]
```diff
@@ -56,14 +65,23 @@ public function verify( array $request ) {
...
 	}
 
-	private function matchesReceiverAddress( $request ): bool {
+	private function matchesReceiverAddress( array $request ): bool {
 		return array_key_exists( 'receiver_email', $request ) &&
 			$request['receiver_email'] === $this->config['account-address'];
 	}
 
-	private function isPaymentStatusAllowed( $request ): bool {
+	private function isPaymentStatusAllowed( array $request ): bool {
```
Comment (JeroenDeDauw) [inline src/Infrastructure/PayPalPaymentNotificationVerifier.php]: bit weird wording: I'd go for something like `paymentStatusIsAllowed` or `isAllowedPaymentStatus`

File: src/UseCases/HandlePayPalPaymentNotification/HandlePayPalPaymentNotificationUseCase.php [hunk 2]
```diff
@@ -4,17 +4,48 @@
...
+	private $authorizationService;
+
+	public function __construct( DonationRepository $repository, DonationAuthorizer $authorizationService ) {
+		$this->repository = $repository;
+		$this->authorizationService = $authorizationService;
 	}
 
-	public function handleNotification( HandlePayPalPaymentNotificationRequest $request ) {
-		return 'TODO';
+	public function handleNotification( PayPalNotificationRequest $request ) {
```
Comment (JeroenDeDauw) [inline src/UseCases/HandlePayPalPaymentNotification/HandlePayPalPaymentNotificationUseCase.php]: : `bool`

Will this stay just a bool?

File: src/Domain/Model/Donation.php [hunk 3]
```diff
@@ -132,16 +132,36 @@ public function cancel() {
...
+		if ( !$this->isPaymentTypeExternal() ) {
+			throw new RuntimeException( 'Only external payments can be confirmed as booked' );
+		}
+
+		if ( !$this->isStatusIncomplete() ) {
+			throw new RuntimeException( 'Only incomplete donations can be confirmed as booked' );
+		}
+
+		$this->status = self::STATUS_EXTERNAL_BOOKED;
+	}
```
Comment (JeroenDeDauw) [inline src/Domain/Model/Donation.php]: I'd have written a small unit test for this

File: src/UseCases/HandlePayPalPaymentNotification/HandlePayPalPaymentNotificationUseCase.php [hunk 4]
```diff
@@ -4,17 +4,48 @@
...
+			return false;
+		}
+
+		if ( $donation === null ) {
+			// TODO: create new donation
+			return false;
+		} else {
+			if ( !$this->authorizationService->canModifyDonation( $request->getDonationId() ) ) {
+				return false;
+			}
```
Comment (JeroenDeDauw) [inline src/UseCases/HandlePayPalPaymentNotification/HandlePayPalPaymentNotificationUseCase.php]: This looks like it is not tested yet

**Comment by JeroenDeDauw** [comment]
Beware that we have several relevant test doubles for this stuff already: failing and succeeding authorizers and `ThrowingEntityManager` for testing db interaction errors

File: src/Infrastructure/PayPalPaymentNotificationVerifier.php (hunk 1 above)
Comment (KaiNissen) [inline src/Infrastructure/PayPalPaymentNotificationVerifier.php]: right. changed it to `hasAllowedPaymentStatus()`.

File: src/UseCases/HandlePayPalPaymentNotification/HandlePayPalPaymentNotificationUseCase.php (hunk 2 above)
Comment (KaiNissen) [inline src/UseCases/HandlePayPalPaymentNotification/HandlePayPalPaymentNotificationUseCase.php]: I think so. PayPal wouldn't process the response body, but only the status code.

File: src/Domain/Model/Donation.php (hunk 3 above)
Comment (KaiNissen) [inline src/Domain/Model/Donation.php]: done

**Pull request merged**